### DISCUSSION TITLE
Proposal: Use GUIDs as IDs instead of extremely long JSON blobs

### DISCUSSION THREAD

**Discussion by nabsul** [dagger/dagger#5136] [Ideas]
A 64KB (or more?) id is really hard to work with. How about assigning a unique GUIDs to steps in Dagger, and referencing them that way? I'm sure the data in the current Base64-encoded JSON blobs is useful, so I would propose to have a way to exchange one for the other. But for referencing purposes, a simple, short GUID would be much more convenient for copy-pasting and storing scenarios.

**Comment by shykes** [comment]
Hi @nabsul, the term "ID" can be confusing (we have discussed changing the name in the past, but no consensus on that at the moment). These are not just references, but really a complete export of the pipeline state. This is a fundamental part of the engine's design. See for example #3923 for a discussion of future applications of this.

It looks like you're looking for a more "regular" ID, something that uniquely identifies a pipeline without actually containing its entire state. That is a perfectly reasonable thing to expect.

Could you share details about your use case, so we can figure out the right design?

**Comment by nabsul** [comment]
One scenario I'm thinking about: I'm not a fan of the hugely nested GraphQL Syntax.

Instead of:

```
withExec(command1) {
   withExec(command2) {
      withExec(command3) {
....
}
```

I would rather be able to do something like (in rough pseudo code):

```
id1 = .... withExec(command1)
id2 = layer(id1) { withExec(command2) }
id3 = layer(id2) { withExec(command3) }
....
```

And of course the above flat list of commands could easily be represented in a for-loop.

**Reply by nabsul** [reply]
Basically the idea is to build the DAG in steps instead of needing to generate the whole thing in one go.

**Comment by gerhard** [comment]
This **may** be related, but the context is different:
* https://github.com/dagger/dagger/pull/5013
